test/cpp/cpp_test_util.cpp

```diff
@@ -304,12 +311,41 @@ void TestBackward(
   torch::Tensor output = testfn(input_vars);
   torch::Tensor xoutput = testfn(xinput_vars);
   AllClose(output, xoutput, rtol, atol);
-  output.backward(torch::ones_like(output));
-  xoutput.backward(torch::ones_like(xoutput));
-  for (size_t i = 0; i < inputs.size(); ++i) {
-    if (inputs[i].defined() && inputs[i].requires_grad()) {
-      ASSERT_TRUE(xinput_vars[i].grad().defined());
-      AllClose(input_vars[i].grad(), xinput_vars[i].grad(), rtol, atol);
+
+  // Calculating double backward requires create_graph=true
+  auto grads = torch::autograd::grad(
+      {output}, inputs_w_grad,
+      /*grad_outputs=*/{torch::ones_like(output)},
+      /*retain_graph=*/c10::nullopt, /*create_graph=*/double_backward);
+  auto xgrads = torch::autograd::grad(
+      {xoutput}, xinputs_w_grad,
+      /*grad_outputs=*/{torch::ones_like(xoutput)},
+      /*retain_graph=*/c10::nullopt, /*create_graph=*/double_backward);
+  for (size_t i = 0; i < grads.size(); ++i) {
+    ASSERT_TRUE(xgrads[i].defined());
+    AllClose(grads[i], xgrads[i], rtol, atol);
+  }
+
+  if (double_backward) {
```

Comment:
Does pytorch now support higher grade derivative?
If yes, I think it would be nicer to accept an argument telling the level of derivative instead of a bool telling if 1 or 2.

Reply:
@davidel Yea pytorch supports higher grade derivative for a subset of ops, unfortunately it's not easy to tell the level of derivative each op supports. For example, composite ops could theoretically support much higher grade derivative while explicit derivative only support first order differentiation. 
Since the most commonly used ones are first and second order derivative, I feel it's fine that we just test up to double backward for now? Wdyt?

Reply:
Isn't that just replacing an `if (double_backward)` with a `for ()` loop and keep updating `grads` and `xgrads`?

Reply:
I see your point now, updated! ;) 
Btw for group_norm/layer_norm we just test again derivative_level=2, but technically their current implementation can support even higher order derivative.